src/Microsoft.Azure.WebJobs.Host/Retry/RetryManager.cs

```diff
@@ -0,0 +1,159 @@
+        public async Task ExecuteWithRetriesAsync(Func<int, object[], Task<IEnumerable<FunctionResult>>> execute, object[] executeParams, CancellationToken token)
+        {
+            await ExecuteWithRetriesInternalAsync(execute, executeParams, _retry, token);
+        }
+
+        private async Task ExecuteWithRetriesInternalAsync(Func<int, object[], Task<IEnumerable<FunctionResult>>> execute, object[] executeParams, RetryResult retry, CancellationToken token)
+        {
+            IEnumerable<FunctionResult> executionResult = null;
+            try
+            { 
+                if (retry != null)
+                {
+                    AsyncFallbackPolicy fallbackPolicy = Policy
+                           .Handle<FunctionInvocationException>()
+                           .FallbackAsync(async (cancellationToken) =>
+                           {
+                               // Adding empty fallback to prevent throwing an exception after all retries have been exhausted 
+                               LogWithId("All retries have been exhausted");
+                               await Task.CompletedTask;
+                           });
+                    Context context = new Context { { "retryCount", 0 }, {"retry", retry} };
+                    await fallbackPolicy
+                        .WrapAsync(GetRetryPolicy(retry))
+                        .ExecuteAsync(async (ctx, tkn) =>
+                        {
+                            int retryCount = (int)ctx["retryCount"];
+                            LogWithId($"Execute with retry, RetryCount: '{retryCount}', Retry: '{retry.Format()}'");
+                            executionResult = await execute(retryCount, executeParams);
+                            RethrowExceptionIfNeeded(executionResult, true);
+                        }, context, token);
+                }
+                else
+                {
+                    LogWithId("Execute without retry");
+                    executionResult = await execute(-1, executeParams);
+                    RethrowExceptionIfNeeded(executionResult, false);
+                }
+            }
+            catch (RetryException retryException)
+            {
+                LogWithId($"Function code returned retry settings: '{retryException.RetryResult.Format()}'");
+                await ExecuteWithRetriesInternalAsync(execute, executeParams, retryException.RetryResult, token);
+            }
+            catch (OperationCanceledException)
+            {
+                LogWithId("Retries were canceled");
+            }
+        }
+
+        private void RethrowExceptionIfNeeded(IEnumerable<FunctionResult> results, bool throwInvocationException)
+        {
+            foreach (var result in results)
+            {
+                if (!result.Succeeded && (result.Exception is RetryException || throwInvocationException))
+                {
+                    throw result.Exception;
```

Comment:
Throw a new exception with this set as the inner exception.

Reply:
We do not want to wrap the exception as we need FunctionInvocationException/ RetryException types for retry policy:

```
AsyncFallbackPolicy fallbackPolicy = Policy                       
     .Handle<FunctionInvocationException>()
...
catch (RetryException retryException)
```